src/BotGroups.pas

```diff
@@ -0,0 +1,967 @@
+  end;
+
+  ForeignProps =
+  record
+    playerPos: TD3DXVector3;
+    playerFegyv: Byte;
+    playerHalott: boolean;
+    playerMatWorld: TD3DXMatrix;
+    allies: array of TD3DXVector3;
+    enemies: array of TD3DXVector3;
+  end;
+
+  State =
+  record          
+    pos: TD3DXVector3;
+    mozgas: TD3DXVector3;
+    lastMozgas: TD3DXVector3;
+    rotateX, rotateY: Single;
+    lastMuks: TMuksoka;
+    targetVec: TD3DXVector3; //target labanal
+    isDead: Integer; //tick num, resets to baseRespawnTime, -1 means alive
+    isMoving: boolean;
+    //
+    actionStart: Cardinal; //tick
+    actionLength: Cardinal; //tick num
+    checkTargets: Cardinal; //tick num, checks on zero, resets to baseReactionTime
+    canShoot: Single; //tick num, resets to baseLovesCooldown
+    aiming: boolean;
+  end;
+
+  TBot = class
+  private
+    ownProps: OwnProps;
+    foreignProps: ForeignProps;
+    state: State;
+    //
+    function hasOjjektumCollide: boolean;
+    procedure resetMozgas;
+    procedure applyMozgas;
+    function findTarget: TD3DXVector3;
+    procedure facePoint(_pos: TD3DXVector3);
+    procedure shootPoint(_pos: TD3DXVector3);
+    function applyAccuracy(_pos: TD3DXVector3): TD3DXVector3; //actually inaccuracy
+    procedure putRagdoll(_pos: TD3DXVector3; forceRajtam: TD3DXVector3);
+  public
+    constructor Create(_id: Integer; _botMode: TBotMode; _fegyv: Byte; _pos: TD3DXVector3);
+    function Draw(mukso: TMuksoka; fegyver: TFegyv; campos: TD3DXVector3): boolean;
+    //
+    function getFegyv: BYTE;
+    function getIsDead: boolean;
+    function getFegyvMatr: TD3DMatrix;
+    //
+    procedure doLogic;
+    procedure passInfo(_playerHalott: boolean; _playerFegyv: Byte; _playerPos: TD3DXVector3; _playerMatWorld: TD3DXMatrix; _bots: array of TBot);
+    procedure collideProjectile(loves: TLoves);
+  end;
+
+  TBotArray = array of TBot;
+
+  TBotGroup = class
+  private
+    id: Integer;
+    size: Integer;
+  public
+    botMode: TBotMode;
+    bots: TBotArray;
+    fegyvs: array of Byte;
+    spawnpos: TD3DXVector3;
+    spawnrad: Integer;
+    maxSize: Integer;
+    team: string;
+    constructor Create(
+      _id: Integer;
+      _maxSize: Integer;
+      _team: string;
+      _botMode: TBotMode;
+      _spawnpos: TD3DXVector3;
+      _spawnrad: Integer
+    );
+    //optional TODO: destructor Destroy();
+    function Add(bot: TBot): Integer;
+    function AddFegyv(fegyv: Byte): Integer;
+    //TODO: function Remove(index: Integer): Boolean;
+    //TODO: function RemoveById(id: Integer): Boolean;
+    //TODO: function findById(id: Integer): Integer; //TODO: move to Utils.pas
+  end;
+
+implementation
+
+
+///////////////////////
+//   HELPERS FUNCS   //
+///////////////////////
+
+procedure _addrongybaba(apos, vpos, gmbvec:TD3DXVector3;fegyv, mlgmb:byte;ID:cardinal;muks:TMuksoka;matWorld:TD3DMatrix);
+var
+  szin:cardinal;
+begin
+  // apos.y:=apos.y+0.5;
+
+  if rbszam >= 50 then exit;
```

Comment:
const rbszammax kellene mert az array 0..50, de sima addrongybaba a mainben meg 20-ig megy.

Reply:
Increased limit in main to 50